pkg/agent/controller/packetsampling/packetsampling_controller.go

```diff
@@ -0,0 +1,651 @@
+
+	var packet, matchPacket *binding.Packet
+	var endpointPackets []binding.Packet
+	var ofPort uint32
+
+	if len(podInterfaces) > 0 {
+		packet, err = c.preparePacket(ps, podInterfaces[0], receiverOnly)
+		if err != nil {
+			return err
+		}
+		ofPort = uint32(podInterfaces[0].OFPort)
+		matchPacket = packet
+		klog.V(2).InfoS("PacketSampling packet:", "packet", *packet)
+		if senderOnly && ps.Spec.Destination.Service != "" {
+			endpointPackets, err = c.genEndpointMatchPackets(ps)
+			if err != nil {
+				return fmt.Errorf("couldn't generate endpoint match packets: %w", err)
+			}
+		}
+	}
+
+	c.runningPacketSamplingsMutex.Lock()
+	psState := packetSamplingState{
+		name: ps.Name, tag: ps.Status.DataplaneTag,
+		receiverOnly: receiverOnly, isSender: isSender,
+		maxNumCapturedPackets: ps.Spec.FirstNSamplingConfig.Number,
+	}
+
+	exists, err := fileExists(string(ps.UID))
+	if err != nil {
+		return fmt.Errorf("couldn't check if the file exists: %w", err)
+
+	}
+	if exists {
+		return fmt.Errorf("packet file already exists. this may be due to an unexpected termination")
+	}
+
+	file, err := createPcapngFile(string(ps.UID))
+	if err != nil {
+		return fmt.Errorf("couldn't craete pcapng file: %w", err)
+	}
+
+	writer, err := pcapgo.NewNgWriter(file, layers.LinkTypeEthernet)
+	if err != nil {
+		return fmt.Errorf("couldn't init pcap writer: %w", err)
+	}
+
+	psState.shouldSyncPackets = len(podInterfaces) > 0
+	psState.uid = string(ps.UID)
+	psState.pcapngFile = file
+	psState.pcapngWriter = writer
+
+	if psState.shouldSyncPackets {
+		psState.updateRateLimiter = rate.NewLimiter(rate.Every(samplingStatusUpdatePeriod), 1)
+	}
+	c.runningPacketSamplings[psState.tag] = &psState
+	c.runningPacketSamplingsMutex.Unlock()
+
+	timeout := ps.Spec.Timeout
+	if timeout == 0 {
+		timeout = crdv1alpha1.DefaultPacketSamplingTimeout
+	}
+	if psState.shouldSyncPackets {
+		klog.V(2).InfoS("installing flow entries for PacketSampling.", "name", ps.Name)
+		err = c.ofClient.InstallPacketSamplingFlows(uint8(psState.tag), senderOnly, receiverOnly, matchPacket, endpointPackets, ofPort, timeout)
+		if err != nil {
+			klog.ErrorS(err, "install flow entries failed.", "name", ps.Name)
+		}
+	}
+	return err
+
+}
+
+func createPcapngFile(uid string) (afero.File, error) {
+	return defaultFS.Create(uidToPath(uid))
+}
+
+func fileExists(uid string) (bool, error) {
+	_, err := defaultFS.Stat(uidToPath(uid))
+	if err == nil {
+		return true, nil
+	} else {
+		if os.IsNotExist(err) {
+			return false, nil
+		} else {
+			return false, err
+		}
+	}
+}
+
+// genEndpointMatchPackets generates match packets(with dest endpoint's ip/port info) besides the normal match packet.
+// these match packets will help the pipeline to capture the pod -> svc traffic.
+// TODO: 1. support name based port name 2. dual-stack support
+func (c *Controller) genEndpointMatchPackets(ps *crdv1alpha1.PacketSampling) ([]binding.Packet, error) {
+	if ps.Spec.Destination.Service != "" {
```

Comment:
You already checked `ps.Spec.Destination.Service` before calling this function, maybe no need to check it again here.

Reply:
updated